### DISCUSSION TITLE
求解答选择器中的 <<  关系操作符用法

### DISCUSSION THREAD

**Discussion by jiuqianyuan** [gkd-kit/gkd#299] [Q&A] [answered]
@lisonge
[A <<(a1,a2,a3,a_n) B : A 是 B 的子孙节点, 并且 A.order 满足 a_m-1, A.order 是深度优先先序遍历的索引 (搭配 quickFind 使用)](https://github.com/gkd-kit/selector/blob/a70d956aaba25d1bd07e872c1030b1ca0d50a027/README.md?plain=1#L108)

操作符`<<` 的这个`order`是什么属性，求解答，试了下好像并不与属性`depth` 等价

可在[此快照](https://i.gkd.li/import/13422363)中举例

**Comment by lisonge** [comment, marked as answer]
深度优先先序遍历参考 [querySelectorAll](https://developer.mozilla.org/zh-CN/docs/Web/API/Document/querySelectorAll) 

任意页面打开浏览器控制台运行 `document.querySelectorAll('*')` 会输出一个类数组， order 就是这个顺序，这个 order 并不重要

---

`<<` 一般情况不怎么使用，主要是 `<<n` 搭配 quickFind 使用

这个快照 <https://i.gkd.li/import/13264779> 就是一个很好的例子

这个目标节点 `[id="com.byted.pangle.m:id/tt_splash_skip_btn"]` 由于不是 quickFind 的，我们只能一个一个手动遍历得到，查询速度就慢

但是其祖先节点 `[id="com.coolapk.market:id/ad_container"]` 是可以 quickFind 的，如果是旧版本的规则最多只能写

```txt
matches:[ '[id="com.coolapk.market:id/ad_container"]' , '[id!=null][id="com.byted.pangle.m:id/tt_splash_skip_btn"]']
```

虽然可以使用 quickFind 但是最后一个选择器仍然需要遍历所有节点，只是减少了无用查询，但是查询速度并没有增加

但是如果使用

```txt
[id="com.byted.pangle.m:id/tt_splash_skip_btn"] <<n [id="com.coolapk.market:id/ad_container"]
```

搭配 quickFind 使用就直接一步到位能判断和查询到目标节点

使用 `<<n` + quickFind 就完全足够了，`<<(1,2,3)` 并不怎么使用

**Reply by jiuqianyuan** [reply]
明白了，
比如` [id="com.byted.pangle.m:id/tt_splash_skip_btn"] <<(a_m-1) [id="com.coolapk.market:id/ad_container"]` ,
GKD会通过[document.querySelectorAll('*')](https://developer.mozilla.org/zh-CN/docs/Web/API/Document/querySelectorAll)获得一个NodeList，里面包含该`[id="com.coolapk.market:id/ad_container"]`下所有节点，
(a_m-1)就是将该NodeList序列化后`[id="com.byted.pangle.m:id/tt_splash_skip_btn"]`的索引坐标

(白话文：展开[id="com.coolapk.market:id/ad_container"]下所有节点，从上往下数的顺序(a_m-1))

另：对操作符`<<`而言， `quickFind`是强制要求吗，还是只是搭配使用可让`quickFind`使用更灵活。如果不是强制要求，建议优化文档描述

**Reply by lisonge** [reply]
> 对操作符<<而言， quickFind是强制要求吗

不是，只是更适合上述场景，后续我修改一下

**Comment by Lin-arm** [comment]
想知道 **选择器1** 里的 `<<n` 换成  `<<6` 会不好一点，即变成 **选择器2** 

1. [@ImageView <4 View[childCount=4] <<n [id="android:id/content"]](https://i.gkd.li/i/14133497?gkd=QEltYWdlVmlldyA8NCBWaWV3W2NoaWxkQ291bnQ9NF0gPDxuIFtpZD0iYW5kcm9pZDppZC9jb250ZW50Il0)
2. [@ImageView <4 View[childCount=4] <<6 [id="android:id/content"]](https://i.gkd.li/i/14133497?gkd=QEltYWdlVmlldyA8NCBWaWV3W2NoaWxkQ291bnQ9NF0gPDw2IFtpZD0iYW5kcm9pZDppZC9jb250ZW50Il0)

虽说只要用了 `<<` ，都会通过类似的 `document.querySelectorAll('*')` 获得一个 `NodeList` ，
但 **选择器1** 先快查找到 `[id="android:id/content"]`   节点后，接下来是不是从 `NodeList` 里的第1条逐条对比到第n条来找 `View[childCount=4]` ？
而 **选择器2** 快查后，就直接拿 `NodeList` 里的第6条来对比？如果第6条对不上就直接跳过了去匹配下一条规则了

我担心的是如果 `.MainActivity` 界面的 `[id="android:id/content"]` 节点下的子孙节点有上千个甚至更多，用 **选择器1** 很耗时从而阻塞其它同界面优先级低的规则

还是说相对于 `querySelectorAll` 的耗时，从 `NodeList` 里的第1条逐条对比到第n条 的耗时几乎能忽略不计？然后就干脆不用 `<<` 而是用：

```ts
@ImageView <4 View < View < View < View < FrameLayout < FrameLayout < [id="android:id/content"]
```

**Reply by lisonge** [reply]
我不太了解，你可以自己测试

**Reply by Lin-arm** [reply]
@AIsouler  对于 `<<` 呢，有没有啥经验教教我

**Reply by AIsouler** [reply]
不是很懂，我一般就写 `[aaa] <<n [bbb] `（可快速查询），我的理解就是先通过根节点快速查询找到`[bbb]`，然后遍历 `[bbb]` 的所有子节点查找 `[aaa]`

**Reply by AIsouler** [reply]
所以如果 `[bbb]` 与根节点之间相差的节点不多的话就没必要这么写，比如 `[aaa] <<n [id="android:id/content"]` 就没必要这么写，因为和 `[aaa] <<n [parent=null]` 差不多

**Reply by Lin-arm** [reply]
遍历，不怕遇到子孙节点很多的情况嘛

<img width="280" height="280" alt="image" src="https://github.com/user-attachments/assets/692bdfec-cb0a-4b76-865e-1c6313716713" />

**Reply by AIsouler** [reply]
很多那也没办法

**Comment by Lin-arm** [comment]
@lisonge 大佬开一下这个讨论呗
